sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -454,6 +493,315 @@ pi_result OCL(piEnqueueMemBufferMap)(
 }
 
+//
+// USM
+//
+
+pi_result OCL(piHostMemAlloc)(void **result_ptr, pi_context context,
+                              pi_usm_mem_properties *properties, size_t size,
+                              pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clHostMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clHostMemAllocINTEL_fn>(
+      context, "clHostMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+pi_result OCL(piDeviceMemAlloc)(void **result_ptr, pi_context context,
+                                pi_device device,
+                                pi_usm_mem_properties *properties, size_t size,
+                                pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clDeviceMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clDeviceMemAllocINTEL_fn>(
+      context, "clDeviceMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context), cast<cl_device_id>(device),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+pi_result OCL(piSharedMemAlloc)(void **result_ptr, pi_context context,
+                                pi_device device,
+                                pi_usm_mem_properties *properties, size_t size,
+                                pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clSharedMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clSharedMemAllocINTEL_fn>(
+      context, "clSharedMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context), cast<cl_device_id>(device),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+pi_result OCL(piMemFree)(pi_context context, void *ptr) {
+
+  pi_result RetVal;
+  clMemFreeINTEL_fn FuncPtr;
+  pi_result Err = getExtFuncFromContext<clMemFreeINTEL_fn>(
+      context, "clMemFreeINTEL", &FuncPtr);
+
+  if (Err != PI_SUCCESS) {
+    RetVal = Err;
+  } else {
+    RetVal = cast<pi_result>(FuncPtr(cast<cl_context>(context), ptr));
+  }
```

Comment:
I think it might look simpler if we remove  `Err`.
```c++
  clMemFreeINTEL_fn FuncPtr;
  pi_result RetVal = getExtFuncFromContext<clMemFreeINTEL_fn>(
      context, "clMemFreeINTEL", &FuncPtr);

  if (RetVal != PI_SUCCESS)
    return RetVal;

  return cast<pi_result>(FuncPtr(cast<cl_context>(context), ptr));
```

I suggest using this pattern with early exit after `getExtFuncFromContext` result check to simplify other methods.

Reply:
Good idea.